File: app/src/main/java/skarlat/dev/ecoproject/adapter/CardsViewAdapter.java
package skarlat.dev.ecoproject.adapter;


import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import skarlat.dev.ecoproject.EcoCard;
import skarlat.dev.ecoproject.R;

/**
 * TODO ("Implement adapter for different viewTypes")
 */
public class CardsViewAdapter extends RecyclerView.Adapter<CardsViewAdapter.ViewHolder>  {
    private LayoutInflater inflater;
    private List<EcoCard> card;

    public CardsViewAdapter(Context context, List<EcoCard> card){
        this.card = card;
        this.inflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public CardsViewAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.card_courses_cards, parent, false);
        return new CardsViewAdapter.ViewHolder(view);
    }
   

    @SuppressLint("ResourceAsColor")
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        EcoCard ecoCard = card.get(position);
        EcoCard.Status status = (EcoCard.Status) ecoCard.getStatus();
        Drawable drawable;
        switch (status){
            case CLOSED:
                holder.header.setVisibility(View.GONE);
                holder.description.setVisibility(View.GONE);
                holder.countOpenCard.setVisibility(View.GONE);
                holder.cardClose.setVisibility(View.VISIBLE);
                drawable = holder.backgroundCard.getBackground();
                drawable.setColorFilter(Color.rgb(213,213,213), PorterDuff.Mode.SRC_OVER);
                holder.cardClose.setText(String.valueOf(position));
                break;
            case OPENED:
                holder.header.setVisibility(View.GONE);
                holder.description.setVisibility(View.GONE);
                holder.countOpenCard.setVisibility(View.GONE);
                holder.cardClose.setVisibility(View.VISIBLE);
                holder.cardClose.setText("");
                drawable = holder.backgroundCard.getBackground();
                drawable.setColorFilter(Color.rgb(248, 242, 232), PorterDuff.Mode.SRC_OVER);
                holder.cardClose.setBackgroundResource(R.drawable.ic_play);
                break;
            case WATCHED:
                holder.header.setVisibility(View.VISIBLE);
                holder.header.setText( ecoCard.getTitle() );
                holder.description.setVisibility(View.VISIBLE);
                holder.description.setText(ecoCard.getDescription());
                drawable = holder.backgroundCard.getBackground();
                drawable.setColorFilter(Color.rgb(248, 242, 232), PorterDuff.Mode.SRC_OVER);
                holder.countOpenCard.setVisibility(View.VISIBLE);
                holder.countOpenCard.setText(String.valueOf(position + 1));
                holder.cardClose.setVisibility(View.GONE);
                break;
            default:
                break;
        }
        holder.backgroundCard.setTag(ecoCard);

        if ( getItemCount() - 1 == position ){ // Latest item
            holder.nextCard.setVisibility(View.GONE);
            LinearLayout.LayoutParams linearParams = new LinearLayout.LayoutParams(
                    new LinearLayout.LayoutParams(
                            LinearLayout.LayoutParams.MATCH_PARENT, 240));
            linearParams.setMargins(0,0,0,16);
            holder.backgroundCard.setLayoutParams(linearParams);
            holder.backgroundCard.requestLayout();

        }
    }

    @Override
    public int getItemCount() {
        return card.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        final TextView header, description, countOpenCard, cardClose;
        final View nextCard;
        final RelativeLayout backgroundCard;
        ViewHolder(View view){
            super(view);
            backgroundCard = view.findViewById(R.id.background_card);
            header = view.findViewById(R.id.title_card);
            description = view.findViewById(R.id.descr_card);
            countOpenCard = view.findViewById(R.id.count_open_card);
            cardClose = view.findViewById(R.id.card_close);
            nextCard = view.findViewById(R.id.nextCard);
        }
    }
}

File: app/src/main/java/skarlat/dev/ecoproject/adapter/CardsViewAdapter.kt
package skarlat.dev.ecoproject.adapter

import android.annotation.SuppressLint
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import skarlat.dev.ecoproject.*
import skarlat.dev.ecoproject.includes.dataclass.Course
import skarlat.dev.ecoproject.includes.dataclass.EcoCard


class CardsViewAdapter(private val cardClickListener: (View) -> Unit) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var cards = arrayListOf<Any>()

    companion object {
        const val COURSE_DESCRIPTION_VIEW_TYPE = R.layout.item_course_info
        const val COURSE_CARD_VIEW_TYPE = R.layout.recycler_item_card
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = inflate(viewType, parent)
        return when (viewType) {
            COURSE_DESCRIPTION_VIEW_TYPE -> CourseDescriptionViewHolder(view)
            else -> CourseCardViewHolder(view)
        }
    }

    fun submitList(card: List<Any>) {
        cards.clear()
        cards.addAll(card)
        notifyDataSetChanged()
    }

    @SuppressLint("ResourceAsColor")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
            COURSE_DESCRIPTION_VIEW_TYPE -> onBindCourseDescription(holder as CourseDescriptionViewHolder, position)
            else -> onBindCourseCard((holder as CourseCardViewHolder), position)
        }
    }

    private fun onBindCourseCard(holderCourseCard: CourseCardViewHolder, position: Int) {
        val ecoCard = cards[position] as EcoCard
        val status = ecoCard.status as EcoCard.Status
        val finished = status == EcoCard.Status.WATCHED
        holderCourseCard.description.show(finished)
        holderCourseCard.title.show(finished)
        holderCourseCard.statusIcon.show(!finished)
        holderCourseCard.numberIcon.visible(finished)
        holderCourseCard.state = status
        if (finished) {
            holderCourseCard.title.text = ecoCard.title
            holderCourseCard.description.text = ecoCard.description
        } else {
            val cardBackgroundId = when (status) {
                EcoCard.Status.CLOSED -> R.color.colorGray
                else -> R.color.cardOpenedBG
            }
            holderCourseCard.cardView.setCardBackgroundColor(holderCourseCard.getColor(cardBackgroundId))
            holderCourseCard.statusIcon.setImageResource(if (status == EcoCard.Status.OPENED) R.drawable.ic_play else R.drawable.ic_lock)
        }
        holderCourseCard.cardView.tag = ecoCard

        val numberDrawableId = holderCourseCard.cardView?.context?.resources?.getIdentifier("ic_card_number_${position}", "drawable", "skarlat.dev.ecoproject")
                ?: R.drawable.ic_card_number
        val numberDrawable = holderCourseCard.cardView?.context?.getDrawable(numberDrawableId)
        numberDrawable?.let { holderCourseCard.numberIcon.setImageDrawable(it) }
        if (finished || status == EcoCard.Status.OPENED) {
            holderCourseCard.cardView.setOnClickListener {
                cardClickListener(it)
            }
        }
    }

    private fun onBindCourseDescription(holder: CourseDescriptionViewHolder, position: Int) {
        val courseDescription = cards[position] as Course
        holder.courseTitle.text = courseDescription.title
        holder.courseDescription.text = courseDescription.fullDescription
        holder.leftCards.text = holder.itemView.context.getQuantityStringZero(R.plurals.left_card_plurals, R.string.no_left_cards, courseDescription.leftCards)
        holder.progressOfCourse.max = courseDescription.countCards
        holder.progressOfCourse.progress = courseDescription.countCards - courseDescription.leftCards
    }

    override fun getItemCount(): Int {
        return cards.size
    }


    inner class CourseCardViewHolder internal constructor(view: View) : RecyclerView.ViewHolder(view) {
        val title: TextView = view.findViewById(R.id.title)
        val description: TextView = view.findViewById(R.id.description)
        val numberIcon: ImageView = view.findViewById(R.id.number)
        val statusIcon: ImageView = view.findViewById(R.id.statusIcon)
        val cardView: CardView = view.findViewById(R.id.cardView)
        var state: EcoCard.Status? = null
    }

    inner class CourseDescriptionViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val courseTitle: TextView = view.findViewById(R.id.curs_title)
        val progressOfCourse: ProgressBar = view.findViewById(R.id.progressOfCourse)
        val leftCards: TextView = view.findViewById(R.id.left_cards)
        val courseDescription: TextView = view.findViewById(R.id.course_desc)
    }

    override fun getItemViewType(position: Int): Int {
        return when {
            cards[position] is EcoCard -> COURSE_CARD_VIEW_TYPE
            else -> COURSE_DESCRIPTION_VIEW_TYPE
        }
    }
}